File: libs/domain/models-rs/src/identity.rs
// libs/domain/models-rs/src/identity.rs
/**
 * =================================================================
 * APARATO: IDENTITY DOMAIN MODELS (V11.0 - ZK READY)
 * CLASIFICACIÓN: DOMAIN ENTITIES (L2)
 * RESPONSABILIDAD: DEFINICIÓN DE IDENTIDADES Y PAYLOADS CIFRADOS
 * ESTADO: GOLD MASTER // NO ABBREVIATIONS
 * =================================================================
 */
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Estructura del Payload Cifrado en el Cliente (AES-256-GCM).
/// Sincronizado con VaultCryptoEngine (L1 - TypeScript).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedIdentityPayload {
    /// Texto cifrado codificado en Base64.
    pub cipher_text_base64: String,
    /// Vector de inicialización único para esta operación.
    pub initialization_vector_base64: String,
    /// Sal de derivación utilizada para la llave maestra.
    pub salt_base64: String,
}

/// Entidad de Identidad Soberana para el acceso a Runtimes Cloud.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Identity {
    pub id: Uuid,
    /// Plataforma de ejecución (ej: "google_colab", "kaggle").
    pub platform: String,
    /// Identificador del operador (Email).
    pub email: String,
    /// Almacenamiento flexible de credenciales (Cifradas o Crudas).
    pub credentials_json: String,
    /// Huella digital del navegador del operador.
    pub user_agent: String,
    pub usage_count: u64,
    pub last_used_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub status: IdentityStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum IdentityStatus {
    Active,
    RateLimited,
    Expired,
    Revoked,
}

/// DTO para la ingesta de identidades desde el Dashboard.
#[derive(Debug, Deserialize)]
pub struct CreateIdentityPayload {
    pub platform: String,
    pub email: String,
    /// Soporta tanto el nuevo EncryptedIdentityPayload como el formato legacy.
    pub cookies: serde_json::Value,
    pub user_agent: String,
}

#[derive(Debug, Deserialize)]
pub struct RevokeIdentityPayload {
    pub email: String,
}

File: libs/domain/api-contracts/src/lib/identity.ts
/**
 * =================================================================
 * APARATO: IDENTITY STRATA CONTRACTS (V15.0 - ZERO-KNOWLEDGE)
 * CLASIFICACIÓN: DOMAIN CONTRACTS (ESTRATO L2)
 * RESPONSABILIDAD: DEFINICIÓN DE CREDENCIALES Y ESTADOS DE ACCESO
 * =================================================================
 */

import { z } from "zod";

/**
 * Estados operativos de una identidad en la Bóveda.
 */
export const IdentityStatusSchema = z.enum([
  "active",
  "ratelimited",
  "expired",
  "revoked",
]);

/** Tipo inferido del estado de identidad */
export type IdentityStatus = z.infer<typeof IdentityStatusSchema>;

/**
 * Esquema de payload cifrado (Sincronizado con VaultCryptoEngine L1).
 */
export const EncryptedIdentityPayloadSchema = z.object({
  cipher_text_base64: z.string(),
  initialization_vector_base64: z.string(),
  salt_base64: z.string(),
});

/** Tipo inferido del payload cifrado */
export type EncryptedIdentityPayload = z.infer<typeof EncryptedIdentityPayloadSchema>;

/**
 * Entidad de Identidad Soberana.
 */
export const IdentitySchema = z.object({
  id: z.string().uuid(),
  platform: z.string(),
  email: z.string().email(),
  usage_count: z.number().int().nonnegative(),
  last_used_at: z.string().datetime().nullable(),
  created_at: z.string().datetime(),
  status: IdentityStatusSchema,
});

/** Tipo inferido de identidad */
export type Identity = z.infer<typeof IdentitySchema>;

/**
 * Esquema de inyección (Payload de subida).
 */
export const IdentityPayloadSchema = z.object({
  platform: z.string(),
  email: z.string().email(),
  cookies: EncryptedIdentityPayloadSchema,
  userAgent: z.string(),
});

/** Tipo inferido del payload de inyección */
export type IdentityPayload = z.infer<typeof IdentityPayloadSchema>;
